mir_eval/transcription.py

```diff
@@ -0,0 +1,243 @@
+  recall, and F-measure of the note transcription, where an estimated note is
+  considered correct if its pitch, onset and (optionally) offset are
+  sufficiently close to a reference note
+
+'''
+
+import numpy as np
+import collections
+from . import util
+import warnings
+
+
+def validate(ref_intervals, ref_pitches, est_intervals, est_pitches):
+    """Checks that the input annotations to a metric look like time intervals
+    and a pitch list, and throws helpful errors if not.
+
+    Parameters
+    ----------
+    ref_intervals : np.ndarray, shape=(n,2)
+        Array of reference notes time intervals (onset and offset times)
+    ref_pitches: list, len=n
+        List of reference pitch values in Hertz
+    est_intervals : np.ndarray, shape=(m,2)
+        Array of estimated notes time intervals (onset and offset times)
+    est_pitches : list, len=m
+        List of estimated pitch values in Hertz
+    """
+    # If reference or estimated notes are empty, warn
+    if ref_intervals.size == 0:
+        warnings.warn("Reference note intervals are empty.")
+    if len(ref_pitches) == 0:
+        warnings.warn("Reference note pitches are empty.")
+    if est_intervals.size == 0:
+        warnings.warn("Estimate note intervals are empty.")
+    if len(est_pitches) == 0:
+        warnings.warn("Estimate note pitches are empty.")
+
+    # Make sure intervals and pitches match in length
+    if not len(ref_intervals) == len(ref_pitches):
+        warnings.warn("Reference intervals and pitches have different "
+                      "lengths.")
+    if not len(est_intervals) == len(est_pitches):
+        warnings.warn("Estimate intervals and pitches have different lengths.")
+
+    # Make sure all pitch values are positive
+    if np.min(ref_pitches) <= 0:
+        warnings.warn("Reference contains at least one non-positive pitch "
+                      "value")
+    if np.min(est_pitches) <= 0:
+        warnings.warn("Estimate contains at least one non-positive pitch "
+                      "value")
+
+
+def precision_recall_f1(ref_intervals, ref_pitches, est_intervals, est_pitches,
+                        onset_tolerance=0.05, pitch_tolerance=50.0,
+                        offset_ratio=0.2, with_offset=False):
+    """Compute the Precision, Recall and F-measure of correct vs incorrectly
+    transcribed notes. "Correctness" is determined based on note onset, pitch
+    and (optionally) offset: an estimated note is assumed correct if its onset
+    is within +-50ms of a ref note and its pitch (F0) is within +- quarter tone
+    (50 cents) of the corresponding reference note. If with_offset is False,
+    note offsets are ignored in the comparison. It with_offset is True,
+    on top of the above requirements, a correct returned note is required to
+    have an offset value within 20% (by default, adjustable via the
+    offset_ratio parameter) of the ref note's duration around the ref note's
+    offset, or within 50ms, whichever is larger.
```

Comment:
I've changed the minimum-offset-tolerance to be user settable in 606fa79 (in fact it turns out no minimum offset tolerance was being used, so this also caught a bug that made it through the tests).

Regarding `with_offset`, I had it the way you propose before, but I actually prefer to have an explicit argument for this (hence `with_offset`). The `evaluate` method always computes the metrics both with and without offsets. If ignoring offsets is made implicit (by `offset_ratio=None`), then when `offset_ratio=None` the two sets of metrics will be identical, even though one is supposedly with offsets and the other without (unless I add some clunky logic into `evaluate`, which I'd rather not).

By having an explicit argument, this difference is made as clear as possible to the user, and since I expect most users to use the `evaluate` method, I think this avoids potential confusion. Long story short, unless there's serious opposition, I'd rather keep it as is.